src/flb_upstream.c

```diff
@@ -617,12 +624,31 @@ struct flb_connection *flb_upstream_conn_get(struct flb_upstream *u)
               "net.connect_timeout        = %i seconds\n"
               "net.source_address         = %s\n"
               "net.keepalive              = %s\n"
-              "net.keepalive_idle_timeout = %i seconds",
+              "net.keepalive_idle_timeout = %i seconds\n"
+              "net.max_connections        = %i",
               u->tcp_host, u->tcp_port,
               u->base.net.connect_timeout,
               u->base.net.source_address ? u->base.net.source_address: "any",
               u->base.net.keepalive ? "enabled": "disabled",
-              u->base.net.keepalive_idle_timeout);
+              u->base.net.keepalive_idle_timeout,
+              u->base.net.max_connections);
+
+
+    /* If the upstream is limited by max connections, check current state */
+    if (u->base.net.max_connections > 0) {
+        flb_stream_acquire_lock(&u->base, FLB_TRUE);
+
+        total_connections  = mk_list_size(&uq->av_queue);
```

Comment:
This operation is somewhat expensive because `mk_list_size` transverses the list to count the elements.

Wouldn't it be better if we maintained a stand alone counter that's incremented by `create_conn` on success and decremented by `prepare_destroy_conn`?

Unless there's something I'm missing we should go through those two functions regardless of how or why the connection is created or destroyed and that wouldn't affect keepalive either.

Reply:
Another point would be having a top level setting that matches the plugin specific one which would give us some additional flexibility when there is more than one output (for those who don't need to keep a throughput balance instances)

Reply:
just to avoid complexity and reduce the potential errors, I choose to just count every time a new connection is needed. Note that from the moment the end user decides to limit the number of connections it's assuming a performance penalty, this will be just a few ms.